### PR TITLE
[Merged by Bors] - bevy_input: Fix process touch event

### PR DESCRIPTION
**Pull request by light4**
# Objective

- `process_touch_event` in `bevy_input` don't update position info. `TouchPhase::Ended` and `TouchPhase::Cancelled` should use the position info from `pressed`. Otherwise, it'll not update. The position info is updated from `TouchPhase::Moved`.

## Solution

- Use updated touch info. 

---

## Changelog

> This section is optional. If this was a trivial fix, or has no externally-visible impact, feel free to skip this section.

- Fixed: bevy_input, fix process touch event, update touch info

### CODE REVIEW

File: crates/bevy_input/src/touch.rs
Comment (MrGVSV) [inline crates/bevy_input/src/touch.rs:207]: Shouldn't we update the touch phase before inserting? (Same on canceled)

Comment (light4) [reply crates/bevy_input/src/touch.rs:207]: The touch info is from `self.pressed` or event, so I think this is fine.
I can change this if you prefer to update it first.

Comment (light4) [reply crates/bevy_input/src/touch.rs:207]: Actually, the touch info should come from `self.pressed`.
If not found in `self.pressed`, there may be some error occurred, set it to default just to get rid of panic.

Comment (MrGVSV) [reply crates/bevy_input/src/touch.rs:207]: > The touch info is from `self.pressed` or event, so I think this is fine.

You mean it already reflects the `Ended`/`Canceled` phase? I haven't used touches in Bevy before but I assume the event returned from `just_released` would be one of those phases.

Comment (light4) [reply crates/bevy_input/src/touch.rs:207]: The info is updated from phase [TouchPhase::Moved](https://github.com/bevyengine/bevy/blob/main/crates/bevy_input/src/touch.rs#L196-L204)

**Comment by alice-i-cecile** [comment]
Requesting review from @HackerFoo, who has by far the most practical experience with Bevy's touch APIs.

File: crates/bevy_input/src/touch.rs
Comment (MrGVSV) [reply crates/bevy_input/src/touch.rs:207]: Oh interesting. See I'd think that one should also update the phase to `Moved`. 🤔

But maybe I'm just mistaken in how the touch system is supposed to work.

**Review by HackerFoo — APPROVED** [review]
I couldn't understand the PR description. It would be good to update this to make this easier for other reviewers.

From looking at the code and test, it looks like this makes sure `.position`/`.force` gets moved to `.prev_position`/`.prev_force` for `TouchPhase::Ended` (the last update in a motion) as it is in other phases. Before this change, the current position would be assigned to both in this phase.

So this change makes sense and is useful to get a correct delta on the last update.

File: crates/bevy_input/src/touch.rs
Comment (HackerFoo) [inline crates/bevy_input/src/touch.rs:350]: It would be good to add an explanation for this assertion.

Comment (light4) [reply crates/bevy_input/src/touch.rs:351]: updated

**Review by alice-i-cecile — APPROVED** [review]
If you're around, could you try to improve the PR description and add comments to the `if let` blocks added? This change is correct, and I won't block on it, but it would be nice to clean it up as we go.

I'll merge this once that's done or next Monday.

**Comment by light4** [comment]
Rebased and add comments to `if let` blocks.
Sorry, I'm not a native English speaker. It'll be great if you can help me to improve the description.

**Pull request closed without merging**
